modules/translation/i18n/i18n.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package i18n
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"code.gitea.io/gitea/modules/log"
+
+	"gopkg.in/ini.v1"
+)
+
+var (
+	ErrLocaleAlreadyExist = errors.New("lang already exists")
+	ErrLocaleAlreadyAdded = errors.New("lang was already added to another store")
+
+	DefaultLocales = NewLocaleStore()
+)
+
+type locale struct {
+	store    *LocaleStore
+	langName string
+	langDesc string
+	messages *ini.File
+}
+
+type LocaleStore struct {
+	// at the moment, all these fields are readonly after initialization
+	langNames   []string
+	langDescs   []string
+	localeMap   map[string]*locale
+	defaultLang string
+}
+
+func NewLocaleStore() *LocaleStore {
+	return &LocaleStore{localeMap: make(map[string]*locale)}
+}
+
+// AddLocale adds locale into the store, and the locale is attached to the store, it can not be added to another one
+func (ls *LocaleStore) AddLocale(lc *locale) error {
+	if _, ok := ls.localeMap[lc.langName]; ok {
+		return ErrLocaleAlreadyExist
+	}
+
+	if lc.store != nil {
+		return ErrLocaleAlreadyAdded
+	}
```

Comment:
```suggestion
	if _, ok := ls.localeMap[lc.langName]; ok || lc.store != nil {
		return ErrLocaleAlreadyExist
	}
```

Reply:
ErrLocaleAlreadyExist is not the same as ErrLocaleAlreadyAdded.
ErrLocaleAlreadyExist means the locale name already exists in current store.
ErrLocaleAlreadyAdded means this locale had been added to other store.

And I made a new commit to remove the `AddLocale` from this PR, it isn't used at the moment, we can add it in future if there is an real requirement. Now `-17,+7` lines.

Reply:
*Woops* Didn't notice the different naming.